File: lib/net/security.ex
defmodule Net.Security do
  def decrypt(iv, ciphertext, tag) do
    key = Application.get_env(:prism, :key)
    if byte_size(key) != 32 do
      {:error, :invalid_key_size, byte_size(key)}
    else
      try do
        case :crypto.crypto_one_time_aead(:aes_256_gcm, key, iv, ciphertext, <<>>, tag, false) do
          :error ->
            {:error, :decryption_failed}
          dec ->
            {:ok, dec}
        end
      catch
        :error, _ ->
          {:error, :decryption_failed}
      end
    end
  end

  def encrypt(data) do
    key = Application.get_env(:prism, :key)
    iv = :crypto.strong_rand_bytes(12)
    {ciphertext, tag} = :crypto.crypto_one_time_aead(:aes_256_gcm, key, iv, data, <<>>, true)
    {iv, ciphertext, tag}
  end
end

File: client/go/security.go
package main

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
)

type security struct {
	key []byte
}

func newSecurity(key string) *security {
	return &security{
		key: []byte(key),
	}
}

func (s *security) encrypt(data []byte) (iv []byte, ciphertext []byte, tag []byte) {
	iv = make([]byte, 12)
	rand.Read(iv)

	block, err := aes.NewCipher(s.key)
	if err != nil {
		panic(err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		panic(err)
	}

	ciphertext = gcm.Seal(nil, iv, data, nil)
	tag = ciphertext[len(ciphertext)-gcm.Overhead():]
	ciphertext = ciphertext[:len(ciphertext)-gcm.Overhead()]

	return iv, ciphertext, tag
}

func (s *security) decrypt(iv, ciphertext, tag []byte) ([]byte, error) {
	block, err := aes.NewCipher(s.key)
	if err != nil {
		return nil, err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}

	full := append(ciphertext, tag...)
	plaintext, err := gcm.Open(nil, iv, full, nil)
	if err != nil {
		return nil, err
	}

	return plaintext, nil
}
